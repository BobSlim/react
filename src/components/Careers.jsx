import { useState } from 'react'
import styles from './Careers.module.css'

export function Careers() {
  const [count, setCount] = useState(0)

  return (
    <>
      <h1>Application</h1>
      <SectionGeneral></SectionGeneral>
      <SectionEducation></SectionEducation>
      <SectionPractical></SectionPractical>
    </>
  )
}

const genericHandleSubmit = (stateSetter) => (e) => {
  e.preventDefault()
  const form = e.target
  const formData = new FormData(form)
  stateSetter(Object.fromEntries(formData.entries()))
  form.reset()
}

function SectionGeneral() {
  const [personalDetails, setPersonalDetails] = useState({})
  const handleSubmit = genericHandleSubmit(setPersonalDetails)

  function DetailDisplay({ fullName, email, tel }) {
    return (
      <ul className={styles.cards}>
        {fullName && <li>Name: {fullName}</li>}
        {email && <li>Email: {email}</li>}
        {tel && <li>Phone: {tel}</li>}
      </ul>
    )
  }
  return (
    <section>
      <h2>Personal Details</h2>
      <DetailDisplay {...personalDetails}></DetailDisplay>
      <form action="" onSubmit={handleSubmit}>
        <label className={styles.label}>Name<input name="fullName" type="text" /></label>
        <label className={styles.label}>Email<input name="email" type="email" /></label>
        <label className={styles.label}>Ph<input name="tel" type="tel" /></label>
        <button type="submit">Submit</button>
      </form>
    </section>
  )
}

function DisplayTable({ tableRows, description = false }) {
  function DisplayRow({ name, title, start, end, des }) {
    return (
      <tr>
        <td>{name}</td>
        <td>{title}</td>
        <td>{start}</td>
        <td>{end}</td>
        {des && <td>{des}</td>}
      </tr>
    )
  }
  console.log(tableRows)
  if (tableRows.length <= 0) {
    return null
  }
  return (
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Title</th>
          <th>Start</th>
          <th>End</th>
          {description && <th>Description</th>}
        </tr>
      </thead>
      <tbody>
        {tableRows.map(row => <DisplayRow {...row} key={`${row.name}${row.title}`}></DisplayRow>)}
      </tbody>
    </table>
  )
}

function SectionEducation() {
  const [educationTable, setEducationTable] = useState([])
  const addEducation = ({ studyName, studyTitle, studyStart, studyEnd }) => {
    setEducationTable(educationTable.concat({
      name: studyName,
      title: studyTitle,
      start: studyStart,
      end: studyEnd,
    }))
  }
  return (
    <section>
      <h2>Education</h2>
      <DisplayTable tableRows={educationTable}></DisplayTable>
      <form action="" onSubmit={genericHandleSubmit(addEducation)}>
        <label className={styles.label}>School Name<input type="text" name="studyName" /></label>
        <label className={styles.label}>Title of Study<input type="text" name="studyTitle" /></label>
        <label className={styles.label}>Study Start<input type="date" name="studyStart" /></label>
        <label className={styles.label}>Study End<input type="date" name="studyEnd" /></label>
        <button type="submit">Add</button>
      </form>
    </section>
  )
}

function SectionPractical() {
  const [practicalTable, setPracticalTable] = useState([])
  const addPractical = ({ workName, workTitle, workStart, workEnd, workDuties }) => {
    setPracticalTable(practicalTable.concat({
      name: workName,
      title: workTitle,
      start: workStart,
      end: workEnd,
      des: workDuties,
    }))
  }

  return (
    <section>
      <h2>Practical Experience</h2>
      <DisplayTable tableRows={practicalTable} description={true}></DisplayTable>
      <form action="" onSubmit={genericHandleSubmit(addPractical)}>
        <label className={styles.label}>Company Name<input type="text" name="workName" /></label>
        <label className={styles.label}>Position Title<input type="text" name="workTitle" /></label>
        <label className={styles.label}>Responsibilities<textarea type="text" name="workDuties" /></label>
        <label className={styles.label}>Study Start<input type="date" name="workStart" /></label>
        <label className={styles.label}>Study End<input type="date" name="workEnd" /></label>
        <button type="submit">Add</button>
      </form>
    </section>
  )
}