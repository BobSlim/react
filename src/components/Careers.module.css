.card {
  padding: 2em;
}

.read-the-docs {
  color: #888;
}

.hiddenEmpty:empty {
  display: none;
}

table {
  border-collapse: collapse;
}

td,
th {
  border: 1px solid white;
  padding: 0rem 1rem;
}

section {
  display: flex;
  flex-direction: column;
  align-items: center;
}

.label {
  display: flex;
  flex-direction: row;
  justify-content: space-between;
}