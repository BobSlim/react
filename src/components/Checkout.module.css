.products{
    border-spacing: 1rem;
    empty-cells: show;
}

.product {
    box-sizing: border-box;
    border: none;
}

.product td {
    padding: 1rem;
}

.product h3 {
    width: auto;
    margin: 0;
}

.product img {
    width: 100%;
    aspect-ratio: 1/1;
    object-fit: scale-down;
    background-color: white;
    border-radius: 2rem;
    max-height: 8rem;
    display: block;
}