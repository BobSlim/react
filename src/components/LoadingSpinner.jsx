import { Icon } from "@iconify/react";
import styles from "./LoadingSpinner.module.css";

export function LoadingSpinner() {
    return (
        <div className={styles.loadingSpinner}>
            <Icon icon={"mdi:react"}></Icon>
        </div>
    )
}