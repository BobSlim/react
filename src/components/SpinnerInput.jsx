import styles from "./SpinnerInput.module.css";

export const SpinnerInput = ({ 
    inputName, 
    value, 
    setValue,
    min = 0,
}) => {
    const handleClick = (number) => (event) => {
        event.preventDefault()
        validSetValue(parseInt(value ? value : 0) + number)
    }

    const validSetValue = (number) => {
        const newValue = parseInt(Math.max(0, number)).toString()
        setValue(newValue)
    }

    return (
        <div className={styles.spinner}>
            <button onClick={handleClick(-1)}>-</button>
            <input
                type="number"
                name={inputName}
                min={min.toString()}
                value={value}
                onChange={e => setValue(e.target.value)}
                onBlur={e => value ? validSetValue(e.target.value) : null} />
            <button onClick={handleClick(+1)}>+</button>
        </div>
    )
}