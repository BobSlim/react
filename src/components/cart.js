import { findIndexById, findObjectById } from "./utils"

export const setItem = (cart, { id, quantity }) => {
    const itemIndex = findIndexById(cart, id)
    if (itemIndex == -1) {
        return [...cart, { id, quantity }]
    }
    const newCart = [...cart]
    newCart[itemIndex].quantity = quantity
    return newCart
}

export const removeItem = (cart, id) =>
    [...cart.toSpliced(findIndexById(cart, id), 1)]

export const getSubTotal = (products) => (cart, id) =>
    findObjectById(cart, id).quantity * findObjectById(products, id).price

export const getCartTotal = (products) => (cart) =>
    cart.map(basket => getSubTotal(products)(cart, basket.id)).reduce((a, x) => a + x, 0)