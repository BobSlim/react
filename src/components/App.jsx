import { Icon } from "@iconify/react"
import { Outlet, Link } from "react-router-dom"
import { setItem, removeItem } from "./cart"
import { useState, useEffect } from "react"
import { getData } from "./utils"

const NavBar = ({children}) =>
    <nav>
        <Link to="/"><button>Home</button></Link>
        <Link to="/careers"><button>Careers</button></Link>
        <Link to="/game"><button>Game</button></Link>
        <Link to="/shop"><button>Shop</button></Link>
        {children}
    </nav>

const CartButton = ({ items = 0 }) =>
    <Link to="/cart">
        <button>
            <Icon icon="mdi:cart"></Icon>
            {items}
        </button>
    </Link>

export const App = () => {
    const [cart, setCart] = useState([])
    const [products, setProducts] = useState([])

    const loadData = async () => {
        const data = await getData('https://fakestoreapi.com/products')
        setProducts(data)
    }

    useEffect(() => {
        loadData();
    }, [])

    return (
        <>
            <NavBar>
                <CartButton items={cart.length == 0 ? 0 : cart.map(basket => basket.quantity).reduce((acc, obj) => parseInt(acc) + parseInt(obj))} />
            </NavBar>
            <Outlet context={[cart, setCart, products]}/>
        </>
    )
}