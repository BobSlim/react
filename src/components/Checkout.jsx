import { useOutletContext } from "react-router-dom"
import { getCartTotal, getSubTotal, removeItem, setItem } from "./cart"
import { findObjectById, displayPrice } from "./utils"
import { LoadingSpinner } from "./LoadingSpinner"
import styles from "./Checkout.module.css"

export const Checkout = () => {
    const [cart, setCart, products] = useOutletContext()

    const removeFromCart = (id) => {
        setCart(removeItem(cart, id))
    }

    return (
        <main>
            <h1>Your Cart</h1>
            {products.length ?
                <section>
                    <table className={styles.products}>
                        <thead>
                            <tr>
                                <th>Image</th>
                                <th>Name</th>
                                <th>Price</th>
                                <th>Quantity</th>
                                <th>Subtotal</th>
                                <th>Button</th>
                            </tr>
                        </thead>
                        <tbody>
                            {cart.map(basket => {
                                const product = findObjectById(products, basket.id)
                                return (
                                    <CartRow {...product} key={basket.id} quantity={basket.quantity}>
                                        <button onClick={() => removeFromCart(basket.id)}>remove</button>
                                    </CartRow>
                                )
                            })}
                        </tbody>
                        <tfoot>
                            <tr>
                                <td></td>
                                <td></td>
                                <td></td>
                                <td>Total:</td>
                                <td>{displayPrice(getCartTotal(products)(cart))}</td>
                                <td><button>Checkout</button></td>
                            </tr>
                        </tfoot>
                    </table>
                </section>
                :
                <LoadingSpinner />
            }
        </main>
    )
}

const CartRow = ({ 
    id = null, 
    title = "NOTITLE", 
    price = 0,
    quantity = 1, 
    image = "",
    children
}) => {
    const subtotal = price * quantity
    return (
        <tr className={styles.product}>
            <td><img src={image} alt={title} /></td>
            <td>{title}</td>
            <td>{displayPrice(price)}</td>
            <td>{quantity}</td>
            <td>{displayPrice(subtotal)}</td>
            <td>{children}</td>
        </tr>
    )
}
