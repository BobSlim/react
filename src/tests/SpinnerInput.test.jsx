import { SpinnerInput } from "@/components/SpinnerInput"
import { fireEvent, render, screen } from '@testing-library/react';
import { beforeEach } from "vitest";
import userEvent from "@testing-library/user-event"
import { useState } from "react";

describe("Spinner", () => {
    let container
    let user
    let input
    let incButton
    let decButton

    const MockParent = () => {
        const [value, setValue] = useState()
        return (
            <SpinnerInput value={value} setValue={setValue}/>
        )
    }

    beforeEach(() => {
        container = render(<MockParent />).container;
        user = userEvent.setup()
        input = screen.getByRole("spinbutton", { name: "" })
        incButton = screen.getByRole("button", { name: "+" })
        decButton = screen.getByRole("button", { name: "-" })
    })
    it("renders", () => {
        expect(container).toMatchSnapshot();
    })
    it("accepts input", async () => {
        await user.type(input, "123")
        expect(input.value).toMatch("123")
    })
    it("resets input to 0", async () => {
        await user.type(input, "-1")
        fireEvent.blur(input)
        expect(input.value).toMatch("0")
    })

    it("increments", async () => {
        await user.click(incButton)
        expect(input.value).toMatch("1")
    })
    it("decrements", async () => {
        await user.type(input, "2")
        await user.click(decButton)
        expect(input.value).toMatch("1")
    })
    it("refuses to decrement below 0", async () => {
        await user.type(input, "1")
        await user.click(decButton)
        await user.click(decButton)
        expect(input.value).toMatch("0")
    })

})