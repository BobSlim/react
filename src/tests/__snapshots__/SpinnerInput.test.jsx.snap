// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Spinner > renders 1`] = `
<div>
  <div
    class="_spinner_15c661"
  >
    <button>
      -
    </button>
    <input
      min="0"
      type="number"
      value=""
    />
    <button>
      +
    </button>
  </div>
</div>
`;
